src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -233,6 +250,13 @@ public void RequestHeartbeat()
             _monitor.RequestHeartbeat();
         }
 
+        private bool ContainsTimeOutException(MongoConnectionException connectionException)
+        {
+            return connectionException.InnerException is IOException ioException &&
+                   ioException.InnerException is SocketException socketException &&
+                   socketException.SocketErrorCode == SocketError.TimedOut;
```

Comment:
verified with a small sample program with the following settings to trigger a timeout exception:

            settings.SocketTimeout = TimeSpan.FromMilliseconds(100);
that a command will throw an exception with this shape from BinaryConnection.ReceiveBuffer

We can also see that the initial isMaster command ends up in this function by tracing the logic from ConnectionInitializer.InitializeConnection, so the exception through by the initial isMaster will also have this shape.

Reply:
Can this logic be moved into `MongoConnectionException` level (like a new property)?

Reply:
The downside of placing it there would be that it'd introduce a new public API we'd need to maintain (internal doesn't suffice unfortunately), and I could see this logic changing as we gain knowledge about the error codes in`SocketException` . Do you think we should still make this part of the public API? Alternatively, we could create a new extension method as an alternative or leave it as is.

Reply:
I think this logic can be used in other driver parts(and that condition is not really obvious at least for me) and it's pretty the same idea as IsNetworkError field as we added before. I would add it on that level, but I don't really insist on it

Reply:
Added!